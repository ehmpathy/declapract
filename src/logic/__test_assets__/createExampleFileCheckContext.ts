import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
