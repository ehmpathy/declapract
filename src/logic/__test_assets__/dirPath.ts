export const testAssetsDirectoryPath = __dirname;
