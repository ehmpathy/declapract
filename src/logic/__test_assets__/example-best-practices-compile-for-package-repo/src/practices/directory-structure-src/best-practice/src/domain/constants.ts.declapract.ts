export const check = { optional: true }; // some projects don't have constants, and thats fine; this shows what naming should look like though
