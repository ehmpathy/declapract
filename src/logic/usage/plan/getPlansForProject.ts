import {
  FileActionPlan,
  FilePracticeEvaluation,
  PracticeDeclaration,
  ProjectVariablesImplementation,
} from '../../../domain';
import { ProjectCheckContext } from '../../../domain/objects/ProjectCheckContext';
import { withDurationReporting } from '../../../utils/wrappers/withDurationReporting';
import { evaluateProjectAgainstPracticeDeclarations } from '../evaluate/evaluateProjectAgainstPracticeDeclarations';
import { getRequiredActionForFile } from './getRequiredActionForFile';

/**
 * get the plans required to make a project follow the declared practices
 */
export const getPlansForProject = withDurationReporting(
  'getPlansForProject',
  async ({
    practices,
    projectRootDirectory,
    projectVariables,
  }: {
    practices: PracticeDeclaration[];
    projectRootDirectory: string;
    projectVariables: ProjectVariablesImplementation;
  }) => {
    // define the project context
    const project = new ProjectCheckContext({
      projectVariables,
      projectPractices: practices.map((practice) => practice.name),
      getProjectRootDirectory: () => projectRootDirectory,
    });

    // evaluate the project against the practices
    const evaluations = await evaluateProjectAgainstPracticeDeclarations({
      practices,
      project,
    });

    // convert each file evaluation in to a plan per file
    const evaluationsPerFile = evaluations.reduce((summary, thisEvaluation) => {
      const currentState = summary[thisEvaluation.path] ?? []; // default to empty array
      return {
        ...summary,
        [thisEvaluation.path]: [...currentState, thisEvaluation],
      }; // append this evaluation
    }, {} as Record<string, FilePracticeEvaluation[]>);

    // compose evaluations into plans
    const plans = Object.entries(evaluationsPerFile).map(
      ([path, evaluationsForFile]) =>
        new FileActionPlan({
          path,
          evaluations: evaluationsForFile,
          action: getRequiredActionForFile({ evaluations: evaluationsForFile }),
        }),
    );

    // return the plans
    return plans;
  },
);
