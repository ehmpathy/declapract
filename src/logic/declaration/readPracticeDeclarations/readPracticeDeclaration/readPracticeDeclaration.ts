import { ProjectCheckDeclaration, PracticeDeclaration, FileCheckPurpose } from '../../../../domain';
import { listPathsInDirectory } from '../../../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../../../UserInputError';
import { getProjectCheckDeclaration } from './getProjectCheckDeclaration';

export const readPracticeDeclaration = async ({ declaredPracticeDirectory }: { declaredPracticeDirectory: string }) => {
  // grab the file paths in this dir
  const paths = await listPathsInDirectory({ directory: declaredPracticeDirectory });

  // grab the name of the practice
  const practiceName = declaredPracticeDirectory.split('/').slice(-1)[0];

  // find the "best-practice" dir, if one was defined
  const bestPracticeProjectCheck: ProjectCheckDeclaration | null = paths.includes('best-practice')
    ? await getProjectCheckDeclaration({
        purpose: FileCheckPurpose.BEST_PRACTICE,
        declaredProjectDirectory: `${declaredPracticeDirectory}/best-practice`,
      })
    : null;

  // find the "bad-practice" dirs, if any were defined
  const badPracticeProjectChecks: ProjectCheckDeclaration[] = await (async () => {
    if (!paths.includes('bad-practices')) return [];
    const badPracticesPaths = await listPathsInDirectory({ directory: `${declaredPracticeDirectory}/bad-practices` });
    return Promise.all(
      badPracticesPaths.map((path) =>
        getProjectCheckDeclaration({
          purpose: FileCheckPurpose.BAD_PRACTICE,
          declaredProjectDirectory: `${declaredPracticeDirectory}/bad-practices/${path}`,
        }),
      ),
    );
  })();

  // make sure that the `/bad-practices` dir was not typo'ed as `/bad-practice` on accident (happened on more than one occasion ðŸ˜…)
  if (paths.includes('bad-practice'))
    throw new UserInputError(
      `a 'bad-practice/' directory was defined for practice '${practiceName}'. this is probably a typo and should have been 'bad-practices/'`,
      { potentialSolution: "rename 'bad-practice/' to 'bad-practices/'" },
    );

  // ensure atleast one of best-practice or bad-practice is defined
  if (!bestPracticeProjectCheck && !badPracticeProjectChecks.length)
    throw new UserInputError(
      `either a 'best-practice' or atleast one of 'bad-practices' must be defined for '${practiceName}'`,
    );

  // return it
  return new PracticeDeclaration({
    name: practiceName,
    bestPractice: bestPracticeProjectCheck,
    badPractices: badPracticeProjectChecks,
  });
};
