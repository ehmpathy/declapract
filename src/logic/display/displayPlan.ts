import chalk from 'chalk';
import indentString from 'indent-string';
import { withNot } from 'simple-type-guards';

import { FileActionPlan, hasFailed, isFixableCheck, RequiredAction } from '../../domain';
import { getColoredPlanTitle } from './color/getColoredPlanTitle';

export const displayPlan = async ({ plan }: { plan: FileActionPlan }) => {
  // print out the title for this file
  const title = getColoredPlanTitle({ plan });
  console.log(`  * ${title}`); // tslint:disable-line: no-console

  // if there is nothing to do, then don't show more details for this file
  if (plan.action === RequiredAction.NO_CHANGE) return;

  // if there was any failed evaluations, display each practice and show each practice that failed and why
  plan.evaluations.filter(hasFailed).forEach((evaluation) => {
    // display the reasons of the practices that failed
    [
      ...evaluation.checked.bestPractice.filter(hasFailed),
      ...evaluation.checked.badPractices.filter(withNot(hasFailed)),
    ].forEach((failedCheck) => {
      const statusToken = isFixableCheck(failedCheck) ? chalk.red('✕') : chalk.red('✕');
      const fixabilityToken = isFixableCheck(failedCheck) ? chalk.gray('(automatically fixable)') : '';
      console.log(indentString(`${statusToken} practice:${failedCheck.practiceRef} ${fixabilityToken}`, 4)); // tslint:disable-line: no-console
      if (failedCheck.reason) console.log(indentString(failedCheck.reason, 6), '\n'); // tslint:disable-line: no-console
    });
  });
};
