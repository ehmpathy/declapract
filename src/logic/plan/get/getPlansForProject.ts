import { FileActionPlan, FilePracticeEvaluation, PracticeDeclaration } from '../../../domain';
import { evaluteProjectAgainstPracticeDeclarations } from './evaluate/evaluateProjectAgainstPracticeDeclarations';
import { getRequiredActionForFile } from './getRequiredActionForFile';

/**
 * get the plans required to make a project follow the declared practices
 */
export const getPlansForProject = async ({
  practices,
  projectRootDirectory,
}: {
  practices: PracticeDeclaration[];
  projectRootDirectory: string;
}) => {
  // evaluate the project against the practices
  const evaluations = await evaluteProjectAgainstPracticeDeclarations({ practices, projectRootDirectory });

  // convert each file evaluation in to a plan per file
  const evaluationsPerFile = evaluations.reduce((summary, thisEvaluation) => {
    const currentState = summary[thisEvaluation.path] ?? []; // default to empty array
    return { ...summary, [thisEvaluation.path]: [...currentState, thisEvaluation] }; // append this evaluation
  }, {} as Record<string, FilePracticeEvaluation[]>);

  // compose evaluations into plans
  const plans = Object.entries(evaluationsPerFile).map(
    ([path, evaluations]) =>
      new FileActionPlan({
        path,
        evaluations,
        action: getRequiredActionForFile({ evaluations }),
      }),
  );

  // return the plans
  return plans;
};
