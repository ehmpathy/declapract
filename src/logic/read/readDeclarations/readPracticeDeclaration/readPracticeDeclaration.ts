import { CheckProjectDeclaration, PracticeDeclaration } from '../../../../domain';
import { listPathsInDirectory } from '../../../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../../../UserInputError';
import { getCheckProjectDeclaration } from './getCheckProjectDeclaration';

export const readPracticeDeclaration = async ({ declaredPracticeDirectory }: { declaredPracticeDirectory: string }) => {
  // grab the file paths in this dir
  const paths = await listPathsInDirectory(declaredPracticeDirectory);

  // grab the name of the practice
  const practiceName = declaredPracticeDirectory.split('/').slice(-1)[0];

  // find the "best-practice" dir, if one was defined
  const bestPracticeProjectCheck: CheckProjectDeclaration | null = paths.includes('best-practice')
    ? await getCheckProjectDeclaration({ declaredProjectDirectory: `${declaredPracticeDirectory}/best-practice` })
    : null;

  // find the "bad-practice" dirs, if any were defined
  const badPracticeProjectChecks: CheckProjectDeclaration[] = await (async () => {
    if (!paths.includes('bad-practices')) return [];
    const badPracticesPaths = await listPathsInDirectory(`${declaredPracticeDirectory}/bad-practices`);
    return Promise.all(
      badPracticesPaths.map((path) =>
        getCheckProjectDeclaration({ declaredProjectDirectory: `${declaredPracticeDirectory}/bad-practices/${path}` }),
      ),
    );
  })();

  // ensure atleast one of best-practice or bad-practice is defined
  if (!bestPracticeProjectCheck && !badPracticeProjectChecks.length)
    throw new UserInputError(
      `either a 'best-practice' or atleast one of 'bad-practices' must be defined for '${practiceName}'`,
    );

  // return it
  return new PracticeDeclaration({
    name: practiceName,
    bestPractice: bestPracticeProjectCheck,
    badPractices: badPracticeProjectChecks,
  });
};
