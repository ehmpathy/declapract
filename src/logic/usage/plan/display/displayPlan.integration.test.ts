import { RequiredAction } from '../../../../domain';
import { log } from '../../../../utils/logger';
import { testAssetsDirectoryPath } from '../../../__test_assets__/dirPath';
import { readPracticeDeclaration } from '../../../declaration/readPracticeDeclarations/readPracticeDeclaration/readPracticeDeclaration';
import { getPlansForProject } from '../getPlansForProject';
import { displayPlan } from './displayPlan';

const logSpy = jest.spyOn(console, 'log').mockImplementation(() => log.debug); // swap to log debug so its not displaying during tests by default

describe('displayPlan', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should show no change plans correctly', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/prettier`,
    });
    const plans = await getPlansForProject({
      projectRootDirectory: `${testAssetsDirectoryPath}/example-project-fails-prettier`,
      practices: [practice],
      projectVariables: {},
    });
    // console.log(JSON.stringify(plans, null, 2));

    // grab the no change plan
    const plan = plans.find(
      (thisPlan) => thisPlan.action === RequiredAction.NO_CHANGE,
    );
    if (!plan) throw new Error('expected to find the plan here');

    // now display the plan
    await displayPlan({ plan });

    // check that it looks right
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(expect.stringContaining('[NO_CHANGE]'));
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('package.json'),
    );
    expect(logSpy.mock.calls[0]).toMatchSnapshot();
  });
  it('should show automatically fixable plans correctly', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/prettier`,
    });
    const plans = await getPlansForProject({
      projectRootDirectory: `${testAssetsDirectoryPath}/example-project-fails-prettier`,
      practices: [practice],
      projectVariables: {},
    });
    // console.log(JSON.stringify(plans, null, 2));

    // grab the no change plan
    const plan = plans.find(
      (thisPlan) => thisPlan.action === RequiredAction.FIX_AUTOMATIC,
    );
    if (!plan) throw new Error('expected to find the plan here');

    // now display the plan
    await displayPlan({ plan });

    // check that it looks right
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('[FIX_AUTOMATIC]'),
    );
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('.prettierignore'),
    );
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
  it('should show manually fixable plans correctly', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times`,
    });
    const plans = await getPlansForProject({
      projectRootDirectory: `${testAssetsDirectoryPath}/example-project-fails-prettier`,
      practices: [practice],
      projectVariables: {},
    });
    // console.log(JSON.stringify(plans, null, 2));

    // grab the no change plan
    const plan = plans.find(
      (thisPlan) => thisPlan.action === RequiredAction.FIX_MANUAL,
    );
    if (!plan) throw new Error('expected to find the plan here');

    // now display the plan
    await displayPlan({ plan });

    // check that it looks right
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('[FIX_MANUAL]'),
    );
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('package.json'),
    );
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
  it('should show failing bad practice correctly', async () => {
    const practices = await Promise.all([
      await readPracticeDeclaration({
        declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times`,
      }),
    ]);
    const plans = await getPlansForProject({
      projectRootDirectory: `${testAssetsDirectoryPath}/example-project-fails-dates-and-times`,
      practices,
      projectVariables: {},
    });
    // console.log(JSON.stringify(plans, null, 2));

    // grab the no change plan
    const plan = plans.find(
      (thisPlan) => thisPlan.action === RequiredAction.FIX_MANUAL,
    );
    if (!plan) throw new Error('expected to find the plan here');

    // now display the plan
    await displayPlan({ plan });

    // check that it looks right
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('[FIX_MANUAL]'),
    );
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('package.json'),
    );
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
  it('should show failing multiple practices on the same file correctly', async () => {
    const practices = await Promise.all([
      await readPracticeDeclaration({
        declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times`,
      }),
      await readPracticeDeclaration({
        declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/prettier`,
      }),
    ]);
    const plans = await getPlansForProject({
      projectRootDirectory: `${testAssetsDirectoryPath}/example-project-fails-dates-and-times`,
      practices,
      projectVariables: {},
    });
    // console.log(JSON.stringify(plans, null, 2));

    // grab the no change plan
    const plan = plans.find(
      (thisPlan) => thisPlan.action === RequiredAction.FIX_MANUAL,
    );
    if (!plan) throw new Error('expected to find the plan here');

    // now display the plan
    await displayPlan({ plan });

    // check that it looks right
    expect(logSpy).toHaveBeenCalled();
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('[FIX_MANUAL]'),
    );
    expect(logSpy).toHaveBeenCalledWith(
      expect.stringContaining('package.json'),
    );
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
  it.todo('should show failing both manually and automatically correctly');
});
