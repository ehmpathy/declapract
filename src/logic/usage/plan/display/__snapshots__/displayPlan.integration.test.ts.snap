// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayPlan should show automatically fixable plans correctly 1`] = `
Array [
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m .prettierignore[22m",
  ],
  Array [
    "    [33mâœ•[39m practice:prettier.best-practice [90m(fix:appliable)[39m",
  ],
  Array [
    "      [32m- Expected toEqual[39m
      [31m+ Received[39m

      [32m- *.md[39m
      [31m+ .js[39m
",
    "
",
  ],
]
`;

exports[`displayPlan should show failing bad practice correctly 1`] = `
Array [
  Array [
    "  * [1m[1m[31m[FIX_MANUAL][39m[1m package.json[22m",
  ],
  Array [
    "    [31mâœ•[39m practice:dates-and-times.bad-practice.moment ",
  ],
]
`;

exports[`displayPlan should show failing multiple practices on the same file correctly 1`] = `
Array [
  Array [
    "  * [1m[1m[31m[FIX_MANUAL][39m[1m package.json[22m",
  ],
  Array [
    "    [31mâœ•[39m practice:dates-and-times.bad-practice.moment ",
  ],
  Array [
    "    [31mâœ•[39m practice:prettier.best-practice ",
  ],
  Array [
    "      [2mexpect([22m[31mreceived[39m[2m).[22mtoMatchObject[2m([22m[32mexpected[39m[2m)[22m

      [32m- Expected  - 5[39m
      [31m+ Received  + 3[39m

      [2m  Object {[22m
      [32m-   \\"devDependencies\\": ObjectContaining {[39m
      [32m-     \\"prettier\\": StringMatching /^(([0-9][0-9]+|[3-9]).([0-9][0-9]+|[0-9]).([0-9][0-9]+|[0-9])|([0-9][0-9]+|[2-9]).([0-9][0-9]+|[1-9]).([0-9][0-9]+|[0-9])|([0-9][0-9]+|[2-9]).([0-9][0-9]+|[0-9]).([0-9][0-9]+|[0-9]))$/,[39m
      [32m-   },[39m
      [32m-   \\"scripts\\": ObjectContaining {[39m
      [32m-     \\"format\\": \\"prettier --write '**/*.ts' --config ./prettier.config.js\\",[39m
      [31m+   \\"dependencies\\": Object {[39m
      [31m+     \\"date-fns\\": \\"8.21.0\\",[39m
      [31m+     \\"moment\\": \\"5.21.0\\",[39m
      [2m    },[22m
      [2m  }[22m",
    "
",
  ],
]
`;

exports[`displayPlan should show manually fixable plans correctly 1`] = `
Array [
  Array [
    "  * [1m[1m[31m[FIX_MANUAL][39m[1m package.json[22m",
  ],
  Array [
    "    [31mâœ•[39m practice:dates-and-times.best-practice ",
  ],
  Array [
    "      [2mexpect([22m[31mreceived[39m[2m).[22mtoEqual[2m([22m[32mexpected[39m[2m) // deep equality[22m

      Expected: [32mObjectContaining {\\"dependencies\\": ObjectContaining {\\"date-fns\\": Any<String>}}[39m
      Received: [31m{\\"devDependencies\\": {\\"prettier\\": \\"8.21.0\\"}, \\"scripts\\": {\\"format\\": \\"prettier --write '**/*.ts' --config ./prettier.config.js\\"}}[39m",
    "
",
  ],
]
`;

exports[`displayPlan should show no change plans correctly 1`] = `
Array [
  "  * [1m[1m[90m[NO_CHANGE][39m[1m package.json[22m",
]
`;
